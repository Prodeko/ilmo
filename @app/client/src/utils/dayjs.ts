import { Locale } from "antd/lib/locale"
import enUS from "antd/lib/locale/en_US"
import fiFI from "antd/lib/locale/fi_FI"
import svSE from "antd/lib/locale/sv_SE"
import dayjs from "dayjs"
import timezone from "dayjs/plugin/timezone"
import utc from "dayjs/plugin/utc"

import "dayjs/locale/fi"
import "dayjs/locale/en"
import "dayjs/locale/sv-fi"

// Required for dayjs timezone features
// https://day.js.org/docs/en/plugin/timezone
dayjs.extend(utc)
dayjs.extend(timezone)

const locales = {
  fi: ["fi", fiFI],
  en: ["en", enUS],
  se: ["se", svSE],
}

export function setLocale(locale: string): Locale {
  const [localeString, antd] = locale ? locales[locale] : [locales["fi"]]
  dayjs.locale(localeString)
  return antd
}
