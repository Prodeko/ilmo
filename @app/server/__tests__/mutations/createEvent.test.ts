import dayjs from "dayjs";
import slugify from "slugify";

import {
  asRoot,
  createEventDataAndLogin,
  deleteTestData,
  runGraphQLQuery,
  sanitize,
  setup,
  teardown,
} from "../helpers";

beforeEach(deleteTestData);
beforeAll(setup);
afterAll(teardown);

const createEventMutation = `
mutation CreateEvent(
  $slug: String!
  $name: JSON!
  $description: JSON!
  $ownerOrganizationId: UUID!
  $categoryId: UUID!
  $isHighlighted: Boolean
  $isDraft: Boolean
  $eventStartTime: Datetime!
  $eventEndTime: Datetime!
  $registrationStartTime: Datetime!
  $registrationEndTime: Datetime!
) {
  createEvent(
    input: {
      event: {
        slug: $slug
        name: $name
        description: $description
        ownerOrganizationId: $ownerOrganizationId
        categoryId: $categoryId
        isHighlighted: $isHighlighted
        isDraft: $isDraft
        eventStartTime: $eventStartTime
        eventEndTime: $eventEndTime
        registrationStartTime: $registrationStartTime
        registrationEndTime: $registrationEndTime
      }
    }
  ) {
    event {
      id
      slug
      name
      description
      ownerOrganizationId
      categoryId
      isHighlighted
      isDraft
      createdBy
      updatedBy
      createdAt
      updatedAt
    }
  }
}`;

describe("CreateEvent", () => {
  it("can create event while logged in", async () => {
    const {
      organization,
      eventCategory,
      session,
    } = await createEventDataAndLogin();

    const day = dayjs("2021-02-20");
    const daySlug = day.format("YYYY-M-D");
    const slug = slugify(`${daySlug}-testitapahtuma`, {
      lower: true,
    });

    await runGraphQLQuery(
      createEventMutation,

      // GraphQL variables:
      {
        slug: slug,
        name: { fi: "Testitapahtuma", en: "Test event" },
        description: { fi: "Testikuvaus", en: "Test description" },
        ownerOrganizationId: organization.id,
        categoryId: eventCategory.id,
        isHighlighted: true,
        isDraft: false,
        eventStartTime: day.add(1, "days").toISOString(),
        eventEndTime: day.add(2, "days").toISOString(),
        registrationStartTime: day.toISOString(),
        registrationEndTime: day.add(7, "hour").toISOString(),
      },

      // Additional props to add to `req` (e.g. `user: {session_id: '...'}`)
      {
        user: { session_id: session.uuid },
      },

      // This function runs all your test assertions:
      async (json, { pgClient }) => {
        expect(json.errors).toBeFalsy();
        expect(json.data).toBeTruthy();

        const event = json.data!.createEvent.event;

        expect(event).toBeTruthy();
        expect(event.ownerOrganizationId).toEqual(organization.id);
        expect(event.categoryId).toEqual(eventCategory.id);

        expect(sanitize(event)).toMatchInlineSnapshot(`
          Object {
            "categoryId": "[id-3]",
            "createdAt": "[timestamp-1]",
            "createdBy": "[id-4]",
            "description": Object {
              "en": "Test description",
              "fi": "Testikuvaus",
            },
            "id": "[id-1]",
            "isDraft": false,
            "isHighlighted": true,
            "name": Object {
              "en": "Test event",
              "fi": "Testitapahtuma",
            },
            "ownerOrganizationId": "[id-2]",
            "slug": "2021-2-20-testitapahtuma",
            "updatedAt": "[timestamp-1]",
            "updatedBy": "[id-5]",
          }
        `);

        const { rows } = await asRoot(pgClient, () =>
          pgClient.query(`SELECT * FROM app_public.events WHERE id = $1`, [
            event.id,
          ])
        );

        if (rows.length !== 1) {
          throw new Error("Event not found!");
        }
        expect(rows[0].id).toEqual(event.id);
      }
    );
  });

  it("can't create quotas while logged out", async () => {
    const { organization, eventCategory } = await createEventDataAndLogin();

    const day = dayjs("2021-02-20");
    const daySlug = day.format("YYYY-M-D");
    const slug = slugify(`${daySlug}-testitapahtuma`, {
      lower: true,
    });

    await runGraphQLQuery(
      createEventMutation,
      // GraphQL variables:
      {
        slug: slug,
        name: { fi: "Testitapahtuma", en: "Test event" },
        description: { fi: "Testikuvaus", en: "Test description" },
        ownerOrganizationId: organization.id,
        categoryId: eventCategory.id,
        isHighlighted: true,
        isDraft: false,
        eventStartTime: day.add(1, "days").toISOString(),
        eventEndTime: day.add(2, "days").toISOString(),
        registrationStartTime: day.toISOString(),
        registrationEndTime: day.add(7, "hour").toISOString(),
      },

      // Additional props to add to `req` (e.g. `user: {session_id: '...'}`)
      {},

      // This function runs all your test assertions:
      async (json) => {
        expect(json.errors).toBeTruthy();

        const message = json.errors![0].message;
        expect(message).toEqual("Permission denied (by RLS)");
      }
    );
  });
});
