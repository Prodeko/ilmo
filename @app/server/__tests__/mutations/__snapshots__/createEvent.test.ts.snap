// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateEvent can create event while logged in 2`] = `
Object {
  "data": Object {
    "createEvent": Object {
      "event": Object {
        "categoryId": "[id-3]",
        "createdAt": "[timestamp-1]",
        "createdBy": "[id-4]",
        "description": Object {
          "en": "Test description",
          "fi": "Testikuvaus",
        },
        "id": "[id-1]",
        "isDraft": false,
        "isHighlighted": true,
        "name": Object {
          "en": "Test event",
          "fi": "Testitapahtuma",
        },
        "ownerOrganizationId": "[id-2]",
        "slug": "2021-2-20-testitapahtuma",
        "updatedAt": "[timestamp-1]",
        "updatedBy": "[id-5]",
      },
    },
  },
}
`;

exports[`CreateEvent can't create quotas while logged out 1`] = `
Object {
  "data": Object {
    "createEvent": null,
  },
  "errors": Array [
    Object {
      "extensions": Object {
        "exception": Object {
          "code": "42501",
          "message": "Permission denied (by RLS)",
          "severity": "ERROR",
        },
      },
      "locations": Array [
        Object {
          "column": 3,
          "line": 15,
        },
      ],
      "message": "Permission denied (by RLS)",
      "path": Array [
        "createEvent",
      ],
    },
  ],
}
`;
